src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,358 @@
+    /// it to express the minimum Antichain (i.e. any future input is valid).
+    ///
+    /// TO resolve this, we designed a migration to instead use the schema
+    /// (Range<Int32>?, Uint64). Range<Int32> still represents Kafka
+    /// partitions, but instead is meant to represent seen Kafka partitions, as
+    /// well as the range of those still unseen, e.g.:
+    ///
+    /// (Bottom, 0) -> 0 // trivially empty sub-zero elements in domain
+    /// [0, 1) -> 12 // partition 0 offset
+    /// [1, Top] -> 0 // all unseen partitions
+    ///
+    /// In addition to migrating Kafka sources to this schema, we also
+    /// introduce the mimum element (i.e. the 0 lsn) for Postgres sources that
+    /// are not new but whose remap shards are empty.
+    pub async fn migrate(
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+        read_metadata: CollectionMetadata,
+        write_shard: ShardId,
+        // additional information to improve logging
+        id: GlobalId,
+        partitioned_source: bool,
+    ) -> Option<()> {
+        let mut migrator = Self::new(
+            persist_clients,
+            read_metadata.clone(),
+            write_shard,
+            id,
+            partitioned_source,
+        )
+        .await
+        .unwrap()?;
+
+        let (migrated_form, upper) = migrator.next().await?;
+
+        info!(
+            "migrating {:?}'s remap shard {:?} to {:?}",
+            id, read_metadata.remap_shard, write_shard
+        );
+
+        migrator
+            .compare_and_append(migrated_form, Antichain::from_elem(T::minimum()), upper)
+            .await
+            .expect("writing to new shard must succeed");
+
+        Some(())
+    }
+
+    /// Generates a new `RemapHandleMigrator` if there is content in
+    /// `read_metadata`'s remap shard, else returns None.
+    pub async fn new(
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+        read_metadata: CollectionMetadata,
+        write_shard: ShardId,
+        // additional information to improve logging
+        id: GlobalId,
+        partitioned_source: bool,
+    ) -> anyhow::Result<Option<Self>> {
+        let mut persist_clients = persist_clients.lock().await;
+        let persist_client = persist_clients
+            .open(read_metadata.persist_location.clone())
+            .await
+            .context("error creating persist client")?;
+        drop(persist_clients);
+
+        let (current_shard_write_handle, current_shard_read_handle) = persist_client
+            .open::<crate::types::sources::data::SourceData, (), T, Diff>(
+                read_metadata.remap_shard,
+                &format!("reclock {}", id),
+            )
+            .await
+            .context("error opening persist shard")?;
+
+        let (since, upper) = (
+            current_shard_read_handle.since(),
+            current_shard_write_handle.upper().clone(),
+        );
+
+        let as_of = match upper.as_option() {
```

Comment:
I believe using a snapshot will give us the full fidelity of updates modulo consolidation; verified this with Dan [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1672244765103129?thread_ts=1672242618.618509&cid=C01CFKM1QRF). Let me know if there's something I'm missing here that prevents us from using `snapshot_and_fetch`. Note that in the code here I've rewritten to use `Subscribe`, but can easily remove it from the code to simply use snapshots.

Reply:
subscribe is fine, the problem is the selection of `as_of` that is passed to the readers. The "modulo consolidation" is the problem, because that is something that loses fidelity. For example if a shard with `since={0}` and `upper={21}` has the following updates:

```
(0, 0, +1)
(0, 10, -1)
(100, 10, +1)
(100, 20, -1)
(150, 20, +1)
```

By requesting a snapshot at `as_of={20}` you'll only get a single update: `(150, 20, 1)`. Everything else will be compacted away. So you must set the `as_of=since` and then keep calling `Subscribe::next` to get the *unconsolidated* (or, more accurately, *uncompacted*) contents of the shard

Reply:
Got it. I didn't realize consolidation included compaction w/r/t timestamps; I thought it only consolidated diffs.